### ISSUE TITLE
should it be clearer when a `type` is generic?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#20473]
Related: #19120 and #19121

Background: this came from my prototyping of "Proposal 4" in https://github.com/chapel-lang/chapel/issues/17122#issuecomment-1190580050 , namely making the `range` type fully generic by removing the defaults for all of its three fields. I started getting errors akin to "cannot create an array of a generic type" in situations where the generic type comes from client code. Given that our compiler's errors are sometimes confusing, which in some cases reflects the difficulty of figuring out where the actual error is, I propose to give users a tool to require non-generic types. It is analogous to an interface in helping make it clear whether the error is on the library side vs. on the application side.

This proposal is that the `type` keyword means "this must be a concrete type" in these cases:
* a `type` alias, ex. `type R = range;`
* a `type` formal, ex. `proc p(type T)`
* a `type` return intent, ex. `proc t() type`

If the user wants to allow generic types in these contexts, they would need to use different syntax, for example a different keyword or `type(?)` :
* `type(?) GR = range(int, ?);`
* `proc p(type(?) T)`
* `proc p() type(?)`

I expect the case "this may be a generic type" to be rare, so there will be few changes to existing codes, if any.

I propose that this alternative syntax accepts concrete types as well, for example to avoid code changes when a generic type becomes concrete due to other code changes.

**Comment by mppf** [comment]
@vasslitvinov - it wasn't quite clear to me on the first read of this issue, but it sounds like you're proposing a change to all `type` variables where the main motivating use case here is for `type` fields. Right?

**Comment by vasslitvinov** [comment]
Correct, if by "type variable" we mean type fields, type aliases, type formals, and type return intent. (Maybe there are other case(s) I am forgetting??)

I propose that the change will require changing existing code (that works correctly) only when the type variable has a generic type.

**Comment by bradcray** [comment]
This also feels somewhat related to https://github.com/chapel-lang/chapel/issues/18214 to me.

I'll note that, in practice, users have wanted to create type aliases such as:

```chapel
record R {
  type t;
  param p: int;
}

type Rint = R(int);
```

before we had working support for such features, so the current syntax was what seemed intuitive to them.

At first blush, I'm not excited by this proposed change, but am also not quite seeing how it would help.  Whether the user declares type aliases with `type` or `type(?)`, the compiler has the same information about whether or not the type is concrete or generic, right?  Can you clarify how having this feature would have helped with the error messages that led you to propose it?  (specifically, it's not obvious to me how it would have been used or helped in the cases you were investigating).

**Comment by vasslitvinov** [comment]
> users have wanted to create type aliases such as: [...] `type Rint = R(int)`

I doubt that it is an indication of their preference of `type` over `type(?)` or something else in this case, that is, when Rint is a generic type.

> how having this feature would have helped with the error messages that led you to propose it?

My scenario was this, not sure it was `set` or another collection:
```chpl
var s = new set(range);
```

Once `range` went from concrete to generic, this became illegal, and I got:
```
error: set element type cannot currently be generic
```

This error is good. Now think about where it comes from:
```chpl
  proc _checkElementType(type t) {
    if isGenericType(t) {
      compilerWarning('creating a set with element type ' + t:string, 2);
      compilerError('set element type cannot currently be generic', 2);
    }
  }
```

That is, the collection author needs to anticipate this kind of issue and react to it proactively.

My proposal is that instead by writing
```chpl
record set {
    type eltType;
    ....
}
```

the "concrete element type" requirement becomes implicit and is checked by the  compiler automatically.

BTW if I remove that check from our `set`, I get an error that obscures the problem:
```
$CHPL_HOME/modules/standard/Set.chpl:107: In function '_isSerializable':
$CHPL_HOME/modules/standard/Set.chpl:110: error: cannot default-initialize a variable with generic type
$CHPL_HOME/modules/standard/Set.chpl:110: note: 'x' has generic type 'range'
$CHPL_HOME/modules/standard/Set.chpl:110: note: cannot find initialization point to split-init this variable
$CHPL_HOME/modules/standard/Set.chpl:111: note: 'x' is used here before it is initialized
```
